internal/controller/object/object.go

```diff
@@ -340,3 +354,160 @@ func setObserved(obj *v1alpha1.Object, observed *unstructured.Unstructured) erro
+			kubeClient = c.client
+		}
+
+		res := &unstructured.Unstructured{}
+		res.SetAPIVersion(ref.FromObject.APIVersion)
+		res.SetKind(ref.FromObject.Kind)
+		// Try to get referenced resource
+		err := kubeClient.Get(ctx, client.ObjectKey{
+			Namespace: ref.FromObject.Namespace,
+			Name:      ref.FromObject.Name,
+		}, res)
+
+		if err != nil {
+			c.logger.Debug(errGetReferencedResource, "error", err)
+			continue
+		}
+
+		// Remove finalizer if exists
+		finalizer := "object/" + string(obj.UID)
+		if meta.FinalizerExists(res, finalizer) {
+			meta.RemoveFinalizer(res, finalizer)
+			if err := kubeClient.Update(ctx, res); err != nil {
+				c.logger.Debug(errFailedToRemoveFinalizer, "error", err)
+				continue
+			}
+		}
+	}
+}
+
+func (c *external) isNotFound(ctx context.Context, obj *v1alpha1.Object, err error) bool {
+	isNotFound := false
+
+	if kerrors.IsNotFound(err) {
+		isNotFound = true
+	} else if meta.WasDeleted(obj) {
+		// If the Object resource was being deleted but the external resource is
+		// not deletable as management policy is specified, we should return the
+		// external resource not found, so that Object can be deleted by managed
+		// resource reconciler. Otherwise, the reconciler will try to delete the
+		// external resource which breaks the management policy.
+		if policy := obj.Spec.ManagementPolicy; policy == v1alpha1.ObserveCreateUpdate || policy == v1alpha1.Observe {
+			c.logger.Debug("Managed resource was deleted but external resource is undeletable.")
+			isNotFound = true
+		}
+	}
+
+	if meta.WasDeleted(obj) && isNotFound {
+		// If the external resource is not found, we should remove the finalizer
+		// that was previously added to the references of this Object, since the
+		// Object, as the dependant resource, was deleted.
+		c.removeReferenceFinalizers(ctx, obj)
+	}
+
+	return isNotFound
+}
+
+func (c *external) handleLastApplied(obj *v1alpha1.Object, last, desired *unstructured.Unstructured) (managed.ExternalObservation, error) {
+	policy := obj.Spec.ManagementPolicy
+	isUpToDate := false
+
+	if policy == v1alpha1.ObserveDelete || policy == v1alpha1.Observe {
+		// Treated as up-to-date to skip last applied annotation update since we
+		// do not create or update the external resource.
+		isUpToDate = true
+	} else if last != nil && equality.Semantic.DeepEqual(last, desired) {
+		// Mark as up-to-date since last is equal to desired
+		isUpToDate = true
+	}
+
+	if isUpToDate {
+		c.logger.Debug("Up to date!")
+
+		// We usually set condition available after create or update is invoked.
+		// However, it will never happen when policy ObserveDelete or Observe is
+		// specified since it does not allow us to create or update the external
+		// resource. In such a case, let's set condition available explicitly if
+		// the resource has been up to date.
+		if policy == v1alpha1.ObserveDelete || policy == v1alpha1.Observe {
+			obj.Status.SetConditions(xpv1.Available())
+		}
```

Comment:
May be we should always set here, for any policy. And remove the ones under Create/Update?

Reply:
That makes perfect sense to me. Code has been updated to reflect this.